src/lib.rs

```diff
@@ -1,236 +1,217 @@
-                timeout: self.timeout
-            },
-        ).map(|vwt| vwt.value.clone())
+  }
+
+  /// Get a value optionally from the cache, if the value is expired this method will return None
+  /// and delete the value lazily from the cache.
+  /// ```
+  /// use simple_cache_rs::SimpleCache;
+  ///
+  /// let mut cache: SimpleCache<i32, String> = SimpleCache::new(None);
+  ///
+  /// cache.get(&1);
+  /// ```
+  pub fn get(&mut self, key: &K) -> Option<V> {
+    let entry = self.hashmap.get(key)?;
+
+    if let Some(timeout) = self.timeout {
+      if entry.insert_time.elapsed() >= timeout {
+        self.delete(key);
+        return None
+      }
     }
 
-    /// Remove an entry from the cache
-    /// ```
-    /// use simple_cache_rs::SimpleCache;
-    ///
-    /// let mut cache: SimpleCache<i32, String> = SimpleCache::new();
-    ///
-    /// cache.insert(1, String::from("a"));
-    /// cache.delete(&1);
-    /// ```
-    pub fn delete(&mut self, key: &K) -> Option<V> {
-        self.h.remove(key).map(|vwt| vwt.value)
+    Some(entry.value.clone())
+  }
+
+  /// Get all keys that are in the cache
+  /// ```
+  /// use simple_cache_rs::SimpleCache;
+  ///
+  /// let mut cache: SimpleCache<i32, String> = SimpleCache::new(None);
+  ///
+  /// cache.keys();
+  /// ```
+  pub fn keys(&self) -> Vec<K> {
+    self.hashmap.keys()
+                .map(|k| k.clone())
+                .collect::<Vec<K>>()
+  }
+
+  /// Get all values that are in the cache
+  /// ```
+  /// use simple_cache_rs::SimpleCache;
+  ///
+  /// let mut cache: SimpleCache<i32, String> = SimpleCache::new(None);
+  ///
+  /// cache.values();
+  /// ```
+  pub fn values(&self) -> Vec<V> {
+    self.hashmap.values()
+                .map(|v| v.value.clone())
+                .collect::<Vec<V>>()
+  }
+
+  /// Insert a batch of items into the cache
+  /// ```
+  /// use simple_cache_rs::SimpleCache;
+  ///
+  /// let mut cache: SimpleCache<i32, String> = SimpleCache::new(None);
+  ///
+  /// let items = vec!((1, String::from("a")), (2, String::from("b")));
+  /// cache.insert_batch(items);
+  /// ```
+  pub fn insert_batch(&mut self, items: Vec<(K, V)>) {
+    let i_now = Instant::now();
+
+    for item in items {
+      self.hashmap.insert(
+                    item.0,
+                    Entry {
+                      value: item.1,
+                      insert_time: i_now
+                  });
     }
+  }
+
+  /// Insert an item into the cache
+  /// ```
+  /// use simple_cache_rs::SimpleCache;
+  ///
+  /// let mut cache: SimpleCache<i32, String> = SimpleCache::new(None);
+  ///
+  /// cache.insert(1, String::from("a"));
+  /// ```
+  pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+    let entry = self.hashmap.insert(
+                              key,
+                              Entry {
+                                value,
+                                insert_time: Instant::now()
+                            })?;
```

Comment:
Same here, I think this returns an Option<V>. I don't think we need to use the ? operator.

Reply:
Same as above